Improve coverage of itemlist_formaction test

tkerdonc opened the pull request:
This adds support for all commands available from the itemlist
formaction to its test class.

While developping these tests, some non testable command were identified, and will require refactoring to be testable : 

- OP_TOGGLEITEMREAD : uninitialized controller attribute, would require controller::run see this [comment](https://github.com/akrennmair/newsbeuter/pull/420#issuecomment-263888458)
- OP_DELETE : as above
- OP_RELOAD : can be processed by the formaction, but will not reload anything, as the controller's feed attribute is empty
- OP_QUIT, OP_NEXTUNREAD, OP_PREVIOUSUNREAD, OP_RANDOMUNREAD, OP_NEXTUNREADFEED, OP_PREVUNREADFEED, OP_NEXTFEED, OP_PREVFEED  : requires a call to view::run in order to populate view::formaction_stack
- OP_MARKFEEDREAD : can be processed by the formaction, but would require the controller's feeds attribute to be initialized in order to test for read status update.
- OP_SEARCH : requires a call to controller::run in order to initialize controller.rsscache
- OP_EDIT_URLS : requires a call to controller::run in order to initialize controller.urlcfg
- OP_SELECTFILTER, OP_SETFILTER, OP_CLEARFILTER : filters cannot be accessed from the text class, they are either parsed at launch from the config file and stored into its filters attribute (but this requires inside configuration of the controller cfgparser), Or directly added to the formaction's matcher item, who then filters the visible items on refresh. In both cases, access to the result of these filtering commands from the test class seems to be compromised.
- OP_SORT, OP_REVSORT : requests an user input in order to determine the sort to use. Implementing an automated mode for this could be a start, however further modifications might be needed in order to assert that the items are actually sorted
- OP_INT_RESIZE : directly calls invalidate

tkerdonc commented:
For now, this only adds OP_OPEN, OP_OPENINBROWSER and OP_OPENBROWSER_AND_MARK.
As discussed in #419, will separate commands between different TEST_CASES in the next push...

Minoru commented:
Do you need my help with fixing the error Travis uncovered? I generally assume you'll look into it yourself, but if you need my assistance, make a comment.

tkerdonc commented:
Thanks for the offer, but I think I got this one, generating the formatted time string from the timestamp using localtime should do it.

Minoru commented:
Skimmed the code; I think I'll wait for you to refactor it into smaller test cases and review then.

GitHub notified me of new commits being pushed, but not if you rebased things, so just keep piling them on; we can rebase everything later.

tkerdonc commented:
Okay, let's see what travis thinks of this fix.
Looks like I won't have time to refactor tonight, so don't wait up. Thanks for the advice !

tkerdonc commented:
Pushed the refactorized tests. I also started on the OP_TOGGLEUNREAD command, but might need some pointers. It is disabled right now, but it otherwise crashes on line 336 of cache.cpp, while trying to lock the mutex.

`
Program received signal SIGSEGV, Segmentation fault.
__GI___pthread_mutex_lock (mutex=0x0) at ../nptl/pthread_mutex_lock.c:66
66      ../nptl/pthread_mutex_lock.c: No such file or directory.
(gdb) bt
#0  __GI___pthread_mutex_lock (mutex=0x0) at ../nptl/pthread_mutex_lock.c:66
#1  0x00000000004f2487 in __gthread_mutex_lock (__mutex=0x10)
    at /usr/include/x86_64-linux-gnu/c++/4.9/bits/gthr-default.h:748
#2  0x00000000004fa6fa in std::mutex::lock (this=0x10) at /usr/include/c++/4.9/mutex:135
#3  0x00000000004fc8fc in std::lock_guard<std::mutex>::lock_guard (this=0x7fffffffbc50, __m=...)
    at /usr/include/c++/4.9/mutex:377
#4  0x00000000004f4dc9 in newsbeuter::cache::mark_item_deleted (this=0x0, guid="", b=false) at src/cache.cpp:336
#5  0x00000000005305de in newsbeuter::controller::mark_deleted (this=0x7fffffffcec0, guid="", b=false)
    at src/controller.cpp:1452
#6  0x0000000000581d13 in newsbeuter::itemlist_formaction::process_operation (this=0x7fffffffcb20,
    op=newsbeuter::OP_TOGGLEITEMREAD, automatic=false, args=0x0) at src/itemlist_formaction.cpp:147
#7  0x0000000000565494 in newsbeuter::formaction::process_op (this=0x7fffffffcb20, op=newsbeuter::OP_TOGGLEITEMREAD,
    automatic=false, args=0x0) at src/formaction.cpp:139
#8  0x0000000000464476 in ____C_A_T_C_H____T_E_S_T____307 () at test/itemlist_formaction.cpp:326
#9  0x00000000004a77dc in Catch::FreeFunctionTestCase::invoke (this=0x907ba0) at test/catch.hpp:6589
#10 0x000000000049363d in Catch::TestCase::invoke (this=0x943f90) at test/catch.hpp:7526
#11 0x00000000004a68b7 in Catch::RunContext::invokeActiveTestCase (this=0x7fffffffdd40) at test/catch.hpp:6165
#12 0x00000000004a657a in Catch::RunContext::runCurrentTest (this=0x7fffffffdd40, redirectedCout="", redirectedCerr="")
    at test/catch.hpp:6136
#13 0x00000000004a4fae in Catch::RunContext::runTest (this=0x7fffffffdd40, testCase=...) at test/catch.hpp:5956
#14 0x0000000000490e25 in Catch::runTests (config=...) at test/catch.hpp:6304
#15 0x00000000004a714f in Catch::Session::run (this=0x7fffffffe0e0) at test/catch.hpp:6412
#16 0x00000000004a7050 in Catch::Session::run (this=0x7fffffffe0e0, argc=1, argv=0x7fffffffe2d8) at test/catch.hpp:6391
#17 0x0000000000499109 in main (argc=1, argv=0x7fffffffe2d8) at test/test.cpp:12
`

I'll keep looking, but suggestions are always appreciated.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -10,27 +10,164 @@
+	cache rsscache(":memory:", &cfg);
+	cfg.set_configvalue("pager", "cat %f > " + pagerfile.getPath());
+
+	std::shared_ptr<rss_feed> feed(new rss_feed(&rsscache));
```
Apparently you wrote the same code in the previous PR and I missed that during review; sorry for that. Good thing the code isn't set in stone :)

Minoru commented:
I find it a bit strange that `__GI___pthread_mutex_lock` is called on `0x0`, but I can't figure out why is that. Let it lay for now.

tkerdonc commented:
Thanks for the review, I'll correct these and see what additional test can be implemented without being bothered by the mutex. Hopefully I'll figure it out eventually.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -336,99 +340,207 @@ TEST_CASE("process_op(OP_TOGGLEITEMREAD)", "[itemlist_formaction]") {
+	}
+
+	SECTION("OP_SHOWURLS empty feed"){
+		REQUIRE_NOTHROW(itemlist.process_op(OP_SHOWURLS));
```
Seems like in this case you're testing on itemlist without feeds, not on itemlist with an empty feed.

tkerdonc replied:
Good catch, the intent was to cover line 189 the section name is indeed wrong.

Minoru replied:
JFTR, coverage is not the best/only/primary/whatever metric: it shows which areas of the code *aren't* tested, but tells us nothing useful of the ones that are. So even if all the lines are covered, try to still think of some properties that the code exhibits that you might test. The simplest example is multiple options: you'll get 100% coverage by testing both separately, but you definitely should write a test where both are combined.

tkerdonc replied:
I totally agree, and as long as I find a return value or modified property I'll try to assert it. But on another hand some code branches will just check some value and abort, and I still think it is interesting to have the test class go through them both to ensure they don't crash and as a way to track dead code.
Just my two cents anyway, If you prefer I'll stick to testable results.

Minoru replied:
> I still think it is interesting to have the test class go through them both to ensure they don't crash and as a way to track dead code.

Ah, that's good then. Keep doing so!

tkerdonc commented on test/itemlist_formaction.cpp:
```diff
@@ -12,6 +12,7 @@ using namespace newsbeuter;
 
 
 TEST_CASE("process_op(OP_OPEN)", "[itemlist_formaction]") {
+	INFO("Opens an article and display its attribute. This is achieved by setting the pager command to echo the article attributes to a temporary file and parsing it.")
```
Thanks for the details, that definitely makes sense, I'll try to respect that in my next modifications.

tkerdonc commented:
Wow, been a bit sidetracked but this is not dead. Adding a test for OP_BOOKMARK. This also modifies the descriptions, but If you feel I missed the mark on that one too don't hesitate to say, I'll get it eventually ^^.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -56,35 +55,35 @@ TEST_CASE("process_op(OP_OPEN)", "[itemlist_formaction]") {
 
 	std::ifstream pagerFileStream (pagerfile.getPath());
-	REQUIRE( std::getline (pagerFileStream,line) );
+	REQUIRE(std::getline (pagerFileStream,line));
```
Just out of curiosity: why did you decide to stop using spaces here?

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -101,8 +100,7 @@ TEST_CASE("process_op(OP_PURGE_DELETED)", "[itemlist_formaction]") {
 
-TEST_CASE("process_op(OP_OPENBROWSER_AND_MARK)", "[itemlist_formaction]") {
-	INFO("Tests item opening in browser by setting the browser option to echo to a temporary file, that is then parsed.");
+TEST_CASE("OP_OPENBROWSER_AND_MARK passes the proper url to the browser and marks read", "[itemlist_formaction]") {
```
You don't need "proper" here: within itemview it's pretty obvious that "the url" is item's URL.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -56,35 +55,35 @@ TEST_CASE("process_op(OP_OPEN)", "[itemlist_formaction]") {
 }
 
-TEST_CASE("process_op(OP_PURGE_DELETED)", "[itemlist_formaction]") {
+TEST_CASE("OP_PURGE_DELETED works properly", "[itemlist_formaction]") {
```
This is of no immediate consequence, as the test is empty at the moment, but a note for the future: I stopped using words "properly", "correctly", "as expected" and so on because they don't explain *what* behaviour is expected.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -206,7 +202,7 @@ TEST_CASE("process_op(OP_OPENALLUNREADINBROWSER)", "[itemlist_formaction]"){
 		openedItemsCount = 0;
 		if (browserFileStream.is_open()) {
-			while ( std::getline (browserFileStream,line) ) {
+			while ( std::getline (browserFileStream,line)) {
```
Hmm, you removed the space before the closing bracket but left one before the open one. This is just arbitrary, please don't do that.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -383,50 +376,88 @@ TEST_CASE("process_op(OP_SHOWURLS)", "[itemlist_formaction]") {
 	}
 
-	SECTION("OP_SHOWURLS default behaviour"){
+	SECTION("default behaviour"){
```
Can you make this more specific? Just like "correctly", this description doesn't really tell anything about the behaviour we expect. And don't forget to mention that this is with built-in URL viewer.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -383,50 +376,88 @@ TEST_CASE("process_op(OP_SHOWURLS)", "[itemlist_formaction]") {
 
-TEST_CASE("process_op(OP_BOOKMARK)", "[itemlist_formaction]") {
-	//NOTIMPL
+TEST_CASE("OP_BOOKMARK works properly", "[itemlist_formaction]") {
```
This, too, should be specific: "OP_BOOKMARK pipes item's URL to bookmark-cmd" or something.

Minoru commented:
> Wow, been a bit sidetracked but this is not dead.

No worries. But be sure to write if you don't have time to work on this any more as I'd rather merge this (maybe with some quick clean-up) than let it rot.

Thanks for working on this!

tkerdonc commented:
Integrating your remarks and adding a test for OP_BOOKMARK. The test is failing right now, according to the documentation only letters upper and lower case are acceptable as flags, however

- "2" or "£" passed as parameter end up in the item's flags.
- """ passed as parameter deletes one of the other flags
- passing all the possible flags at once causes the capital letters from G to L to be ignored

Let me know if these are aspects your think are interesting to test. If so I can create issues and dig deeper tomorrow.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -418,7 +418,7 @@ TEST_CASE("OP_SHOWURLS properly shows the article's properties", "[itemlist_form
 }
 
-TEST_CASE("OP_BOOKMARK works properly", "[itemlist_formaction]") {
+TEST_CASE("OP_BOOKMARK pipes articles url and title to bookmark-command", "[itemlist_formaction]") {
```
1. "article's", not "articles";
2. "bookmark-cmd", not "bookmark-command".

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -460,8 +460,80 @@ TEST_CASE("OP_BOOKMARK works properly", "[itemlist_formaction]") {
 
-TEST_CASE("process_op(OP_EDITFLAGS)", "[itemlist_formaction]") {
-	//NOTIMPL
+TEST_CASE("OP_EDITFLAGS arguments are added to an item's flags", "[itemlist_formaction]") {
```
The flags aren't *added*—they're *set*, i.e. the new set of flags completely replaces the old one.

Come to think of it, there probably should be a test case demonstrating that flags are overwritten, not just appended or something. Can you please add one?

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -460,8 +460,80 @@ TEST_CASE("OP_BOOKMARK works properly", "[itemlist_formaction]") {
+			REQUIRE(item->flags() == flags);
+		}
+		SECTION("Symbols"){
+			op_args.push_back(flags + "%^\\*;\'\"&~#{([-|`_/@)]=}$£€µ,;:!?./§");
```
"Symbol" is a vague word; I think you should use `LOG` here to display what symbols you add—it will help understand what's wrong if the test fails.

Minoru commented:
> "2" or "£" passed as parameter end up in the item's flags

There's bound to be a user who came to depend on this bug (just as [XKCD describes](https://xkcd.com/1172/)), so we'll have to turn that into a feature, i.e. document it and pretend it's how we wanted it to be :) You can dig into that and create a separate PR with an update to the docs; or if you don't want to work on that at the moment, at least file an issue describing this so someone else can pick it up.

> * """ passed as parameter deletes one of the other flags
> * passing all the possible flags at once causes the capital letters from G to L to be ignored

These are legitimate bugs. You should:
1. file issues about them—it's easier to reference an issue than some obscure comment in the middle of PR's discussion;
2. if you're willing to fix those, do so and submit separate PRs. These PRs should probably contain some tests demonstrating the problem (then we'll be sure you've really fixed it). I'll merge those and you'll rebase this PR onto `master`, merging those tests into what you're working here. It's a bit roundabout, but it splits the work in independent parts where you can step away from any part at any moment and we don't end up with entangled mess of PRs and discussions.
3. if you don't want to work on these bugs, just leave the issues for someone else to pick up.

And last but not least—congratulations on good catch! :)

tkerdonc commented:
Thanks ! I created issues #436 #437 and #438, hopefully I'll have time this weekend to have a go at them.

Minoru commented:
I finally managed to find out why [`TOGGLEITEMREAD` test was failing](https://github.com/akrennmair/newsbeuter/pull/420#issuecomment-260314473) (and I now wonder why I didn't see it right away). The key is fourth frame, where `this` (which should point to `cache` object) is `0x0`. This happens because the code doesn't look at the `cache` inside `rss_feed` or `rss_item`—it's assumed that these were initialized from `controller`, and the call is made directly to the object inside `controller`. And this object isn't initialized because we didn't call `controller::run`.

This is a deficiency of the architecture, and fixing it is clearly out of scope of this PR. Please do the following:
- remove the test;
- start a list in your original post so we can track troubles like these (I suspect we'll see a few more). When we're done with this PR, we'll take this list and make an issue of it, or maybe find some other use of it—we'll see then.

tkerdonc commented:
Cool, thanks for looking into this. Editing the original post, and will remove the test in the next push.

tkerdonc commented:
Rebased to upstream in order to benefit from tag cleaning fixes made in #440 
Tests will still fail because of a typo, will be fixed in next push.

tkerdonc commented:
Added a test for the OP_SAVE command.
Hmmm, starting to have a lot of duplicate code regarding item attributes initialization or output file parsing. might be worth refactoring don't you think ?

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -536,8 +503,71 @@ TEST_CASE("OP_EDITFLAGS arguments are added to an item's flags", "[itemlist_form
-	//NOTIMPL
+TEST_CASE("OP_SAVE writes an article's attributes to the specified file", "[itemlist_formaction]") {
+	controller c;
+	newsbeuter::view v(&c);
```
Funny it's only now I notice this: why is `view` here qualified? We have `using namespace newsbeuter` at the beginning of the file, `view v(&c)` should work just fine.

tkerdonc replied:
Good point, messed it up in the first test, then propagated to all the other ones. Will fix.

tkerdonc replied:
Turns out it was necessary after all, to avoid confusion with the view class in colormanager.h, so I left it as is.

Minoru commented:
> Hmmm, starting to have a lot of duplicate code regarding item attributes initialization or output file parsing. might be worth refactoring don't you think ?

I… frankly don't know.

I'm definitely against pulling *all* the data we use in tests into one setup function. It feels imperative to me to have concise yet self-sufficient tests, and doing as little initialization as possible stems directly from that.

On the other hand, if we approach this conservatively and only factor *identical* stuff out, we won't reap much benefit as not that many tests depend on same exact data.

That's why I've been postponing this refactor: I want to have as big picture as possible before doing anything about it. Maybe we just have bad, untestable architecture? I don't know. At least in its current state the tests are independent of each other and can be handled quite freely; once we entangle them with common setup functions, I'm afraid, making changes will become harder.

Do you have any thoughts yourself? I'd welcome some input here.

tkerdonc commented:
I am totally for leaving test classes independent from one another, as you say, makes for easier writing. I was thinking of grouping some things inside the itemlist_formaction tests, i.e., a function that creates a feed containing an item with all its attributes filled, and another one that reads a file and parses the appropriate feed/item attributes. 
But for now, It would only be helpful in three tests, not that much of a bargain indeed. Let's wait for more to be written.

tkerdonc commented:
Good evening and happy new year. This last push adds a test for th OP_HELP command to the itemlist_formaction test class. It does not test match for now, so keeping it can be discussed. On the other hand, implementing and automatic mode (such as OP_SAVE does) could be usefull for testing.

Minoru commented:
Happy new year! The code looks good; glad you're still working on this.

tkerdonc commented:
Committing a test for the OP_RELOAD command, but it is not working right now. We would require the controller's feed attribute in order to actually do a feed reload. Unless you see a better solution, we can either : 

- Completely remove the test
- Leave the call to "REQUIRE_NOTHROW(itemlist.process_op(OP_RELOAD));" in order to at least ensure the command can be sent, but not test for an actual update.

Also adding OP_RELOAD to the top post, at least for now

Minoru commented:
I don't see an easy enough way to use in test; so just remove it.

tkerdonc commented:
Agreed. 
Had a look a tests for OP_QUIT, OP_NEXTUNREAD and OP_PREVIOUSUNREAD today, unfortunately, they would require a call to view::run in order to populate view::fomaction_stack, adding to the first post.

tkerdonc commented:
Added a test for OP_PREV and OP_NEXT commands. Originally wanted to include more elaborate tests with feed switching, or cases where there is no actual next or previous item to go to, but those would crash as they require view::feedlist, which is initialized by view::run.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -597,66 +599,80 @@ TEST_CASE("OP_HELP command is processed", "[itemlist_formaction]") {
-	//NOTIMPL
-}
+TEST_CASE("Navigate back and forth using OP_NEXT and OP_PREVIOUS", "[itemlist_formaction]") {
+	INFO("We are using the OP_SHOWURLS command to print the current article'attibutes to a file, and assert the position was indeed updated.");
```
I think this belongs to a simple comment, not `INFO`. The person for whom this test will fail probably won't need this information to find out what's broken.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -597,66 +599,80 @@ TEST_CASE("OP_HELP command is processed", "[itemlist_formaction]") {
 	keymap k(KM_NEWSBEUTER);
 	v.set_keymap(&k);
-
+	
```
Ooops, trailing whitespace!

Minoru commented:
> Hmmm, starting to have a lot of duplicate code regarding item attributes initialization or output file parsing. might be worth refactoring don't you think ?

On a second thought, output file parsing is indeed a good candidate for factoring out: it doesn't impose anything on us, and it doesn't entangle tests. I suspect we'll also need this in other formaction tests, so you can put it in `test/test-helpers.h`.

tkerdonc commented:
Pushing an update with the following changes : 
- proposal for a file parsing function in test-helpers.h
- added a test for the OP_TOGGELSHOWREAD command

Also added some non testable commands to the top post, for already encountered reasons : 
- OP_RANDOMUNREAD, OP_NEXTUNREADFEED, OP_PREVUNREADFEED, OP_NEXTFEED, OP_PREVFEED : requires a call to view::run in order to populate view::formaction_stack
- OP_MARKFEEDREAD : can be processed by the formaction, but would require the controller's feeds attribute to be initialized in order to test for read status update.

Minoru commented on test/test-helpers.h:
```diff
@@ -113,6 +114,42 @@ namespace TestHelpers {
+	 * appropriate "pager" config value ) and checks the differents lines according
+	 * to the expected values passed as parameters.
+	 */
+	inline void AssertArticleFileContent( std::string path, std::string title, std::string author, std::string date, std::string url, std::string description) {
```
Since `inline` doesn't necessarily mean that the function will be inlined, it's better to declare arguments as `const std::string&`.

Minoru commented on test/test-helpers.h:
```diff
@@ -113,6 +114,42 @@ namespace TestHelpers {
+	/*
+	 * The AssertArticleFileContent opens a file where the content of an article
+	 * was previously dumped (using for example OP_SHOWURL, or OP_OPEN with an
+	 * appropriate "pager" config value ) and checks the differents lines according
```
"checks the differents lines" is obviously a typo of some sort.

Minoru commented on test/itemlist_formaction.cpp:
```diff
@@ -675,32 +616,39 @@ TEST_CASE("Navigate back and forth using OP_NEXT and OP_PREVIOUS", "[itemlist_fo
+	SECTION("True to False"){
+		v.get_cfg()->set_configvalue("show-read-articles", "yes");
+		REQUIRE_NOTHROW(itemlist.process_op(OP_TOGGLESHOWREAD));
+		REQUIRE(!v.get_cfg()->get_configvalue_as_bool("show-read-articles"));
```
Catch has `REQUIRE_FALSE` macro which I like because it's much more visible than negation operator. Please use it here.

Minoru commented:
You rock!

tkerdonc commented:
Thanks ! The last push integrates the changes you requested and add a test for the OP_PIPE_TO command. Also adding two problematic commands to the top post (OP_EDIT_URLS & OP_SEARCH).

Minoru commented on test/test-helpers.h:
```diff
@@ -118,10 +118,10 @@ namespace TestHelpers {
 	 * to the expected values passed as parameters.
 	 */
-	inline void AssertArticleFileContent( std::string path, std::string title, std::string author, std::string date, std::string url, std::string description) {
+	inline void AssertArticleFileContent( const std::string path, const std::string title, const std::string author, const std::string date, const std::string url, const std::string description) {
```
`const std::string` means we want a *copy* of a string, but we promise not to modify it. But if we aren't modifying it, why request a copy? We could just take a pointer or, safer yet, a reference to it. That's why I mentioned `const std::string&` in my previous review. Please use references here.

tkerdonc replied:
As always, you detailed instructions are very appreciated, the last push fixes this.

tkerdonc commented:
Wow, sorry for the delay, real life got in the way, but this last push concludes the list of OPerations passable to the process_op function. In a pretty disappointing way though, since I could not manage to test the last ones. Details are as follow : 

- OP_SELECTFILTER, OP_SETFILTER, OP_CLEARFILTER : filters cannot be accessed from the text class, they are either parsed at launch from the config file and stored into its filters attribute (but this requires inside configuration of the controller cfgparser), Or directly added to the formaction's matcher item, who then filters the visible items on refresh. In both cases, access to the result of these filtering commands from the test class seems to be compromised.
- OP_SORT, OP_REVSORT : requests an user input in order to determine the sort to use. Implementing an automated mode for this could be a start, however further modifications might be needed in order to assert that the items are actually sorted
- OP_INT_RESIZE : directly calls invalidate

Minoru commented:
Increasing the coverage by more than four percent is no small feat, so don't feel upset if a few functions didn't lend themselves to testing ;) Thank you for your work! I'll sort out the list in the top posting shortly.

tkerdonc commented:
My pleasure, thanks for the continuous support and the final merging. I am subscribing to the issues you created. Not sure, but maybe the decoupling would also help with the automated testing of the feedlist_formaction...
